### PR TITLE
Improve test coverage of Navier-Stokes module

### PR DESCRIPTION
**Pull request by jwpeterson**
This PR improves (increases from 0%) the test coverage of the compressible Navier-Stokes kernels in the navier_stokes module.  The new test is based on meshes from @yidongxiainl for a well-known benchmark problem for which we can estimate the error in the L2 norm.  

I will also be converting this module to use @andrsd's new `FluidProperties` stuff in the near future, but we wanted to get test coverage in place before undertaking that task.

Refs #3036.

### CODE REVIEW

File: modules/navier_stokes/include/auxkernels/NSMachAux.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * Auxiliary kernel for computing the Mach number assuming an ideal gas.
+ */
+class NSMachAux : public AuxKernel
+{
+public:
+  /**
+   * Constructor.
+   */
```
Comment (andrsd) [inline modules/navier_stokes/include/auxkernels/NSMachAux.h]: cpt. Obvious

File: modules/navier_stokes/src/auxkernels/NSMachAux.C [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+#include "NSMachAux.h"
+#include "MooseMesh.h"
+
+template<>
+InputParameters validParams<NSMachAux>()
+{
+  InputParameters params = validParams<AuxKernel>();
+
+  // Mark variables as required
```
Comment (andrsd) [inline modules/navier_stokes/src/auxkernels/NSMachAux.C]: Some are required, some are not. We can see which ones from the name. Also, for 1D run, `y-velocity` should not be required.

File: modules/navier_stokes/src/auxkernels/NSMachAux.C [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  params.addRequiredCoupledVar("u", "x-velocity");
+  params.addRequiredCoupledVar("v", "y-velocity");
+  params.addCoupledVar("w", "z-velocity"); // Only required in 3D...
+  params.addRequiredCoupledVar("temperature", "");
+
+  // Parameters with default values
```
Comment (andrsd) [inline modules/navier_stokes/src/auxkernels/NSMachAux.C]: I do not see any default values here...

File: modules/navier_stokes/src/auxkernels/NSMachAux.C [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  params.addRequiredParam<Real>("R", "Gas constant.");
+
+  return params;
+}
+
+NSMachAux::NSMachAux(const InputParameters & parameters) :
+    AuxKernel(parameters),
+    _u_vel(coupledValue("u")),
+    _v_vel(coupledValue("v")),
+    _w_vel(_mesh.dimension() == 3 ? coupledValue("w") : _zero),
```
Comment (andrsd) [inline modules/navier_stokes/src/auxkernels/NSMachAux.C]: I think, we should be actually doing: `isCoupled("w") ? coupledValue("w") : _zero`, so we can run lower order elements in higher dimensional spaces.  I know we have this all over the NS module. I'd like to abandon the `_mesh.dimension()` check and go with `isCoupled()`.  Unless there is a better way how to do it...

File: modules/navier_stokes/src/bcs/NSPressureNeumannBC.C [hunk 5]
```diff
@@ -4,70 +4,63 @@
...
-// }
+
+#include "NSPressureNeumannBC.h"
+
+template<>
+InputParameters validParams<NSPressureNeumannBC>()
+{
+  InputParameters params = validParams<NSIntegratedBC>();
+
+  // Required vars
```
Comment (andrsd) [inline modules/navier_stokes/src/bcs/NSPressureNeumannBC.C]: cpt. Obvious.

File: modules/navier_stokes/src/bcs/NSPressureNeumannBC.C [hunk 6]
```diff
@@ -4,70 +4,63 @@
...
+  params.addRequiredCoupledVar("pressure", "The current value of the pressure");
+
+  // Required parameters
```
Comment (andrsd) [inline modules/navier_stokes/src/bcs/NSPressureNeumannBC.C]: cpt. Obvious

File: modules/navier_stokes/src/bcs/NSPressureNeumannBC.C [hunk 7]
```diff
@@ -4,70 +4,63 @@
...
+    _component(getParam<unsigned>("component")),
+    _gamma(getParam<Real>("gamma")),
+    // Pressure derivative computation object
+    _pressure_derivs(*this)
+{
+}
+
+
+
+Real NSPressureNeumannBC::computeQpResidual()
```
Comment (andrsd) [inline modules/navier_stokes/src/bcs/NSPressureNeumannBC.C]: return type on a separate line, the same applied below

File: modules/navier_stokes/src/bcs/NSPressureNeumannBC.C [hunk 8]
```diff
@@ -4,70 +4,63 @@
...
+{
+  return _pressure[_qp] * _normals[_qp](_component) * _test[_i][_qp];
+}
+
+
+
+Real NSPressureNeumannBC::computeQpJacobian()
+{
+  return computeJacobianHelper(_component+1);  // <-- the on-diagonal variable number is _component+1
```
Comment (andrsd) [inline modules/navier_stokes/src/bcs/NSPressureNeumannBC.C]: spaces around `+`.

File: modules/navier_stokes/src/auxkernels/NSMachAux.C (hunk 4 above)
Comment (jwpeterson) [inline modules/navier_stokes/src/auxkernels/NSMachAux.C]: I agree we should get rid of `_mesh.dimension()` checks, if you go with `isCoupled()` and somebody forgets to couple something in 3D, then you are going to get silently wrong results.

Comment (andrsd) [inline modules/navier_stokes/src/auxkernels/NSMachAux.C]: > if you go with isCoupled() and somebody forgets to couple something in 3D, then you are going to get silently wrong results.

I know, unfortunately we do not have a better way how to guard against it ;-( we would need a map from block ID -> block dimension in something like `FEProblem` and then use that info in integrity checks. Or something along those lines...

Comment (jwpeterson) [inline modules/navier_stokes/src/auxkernels/NSMachAux.C]: @andrsd and I discussed this, and I think in principle these Kernels could be run on lower-dimensional manifolds living in 3D space, possibly with some slight modifications.  Until we make those, however, the `_mesh.dimension()` check is fast and a bit safer than `isCoupled()`.  I do need to add the same thing for the y-component of velocity, however.

File: modules/navier_stokes/src/kernels/NSTemperatureL2.C [hunk 9]
```diff
@@ -59,16 +59,16 @@ NSTemperatureL2::computeQpJacobian()
 Real
 NSTemperatureL2::computeQpOffDiagJacobian(unsigned int jvar)
 {
-  if (jvar == _p_var)
+  if (jvar == _rho_var)
```
Comment (andrsd) [inline modules/navier_stokes/src/kernels/NSTemperatureL2.C]: 👍

Comment (jwpeterson) [inline modules/navier_stokes/src/kernels/NSTemperatureL2.C]: Heheh, I thought I had fixed all these *many* years ago but one slipped through the cracks.

**Comment by jwpeterson** [comment]
bump_ss_ref5

Here's the test problem's solution on the finest grid.

**Pull request merged**